File: packages/backend/src/source/handler/markdown.handler.ts
import { Injectable, Logger } from '@nestjs/common';
import { ContentHandler } from './content-handler.interface';
import { VectorSaver } from './vector-saver';

export type ChunkStrategy = 'fixed' | 'sentence';

export interface ChunkOptions {
  strategy?: ChunkStrategy;
  chunkSize?: number;
  chunkOverlap?: number;
}

@Injectable()
export class MarkdownHandler implements ContentHandler {
  readonly supportedMimeTypes = [
    'text/plain',
    'text/markdown',
    'application/octet-stream',
  ];
  private readonly logger = new Logger(MarkdownHandler.name);

  constructor(private readonly vectorSaver: VectorSaver) {}

  segmentText(text: string, opts: ChunkOptions = {}): string[] {
    const { strategy = 'fixed', chunkSize = 1000, chunkOverlap = 0 } = opts;

    if (strategy === 'fixed') {
      const chunks: string[] = [];
      for (let i = 0; i < text.length; i += chunkSize - chunkOverlap) {
        chunks.push(text.slice(i, i + chunkSize));
      }
      return chunks;
    }

    const sentences = text.match(/[^.!?]+[.!?]+/g) || [text];
    const chunks: string[] = [];
    let current = '';

    for (const sentence of sentences) {
      if ((current + sentence).length > chunkSize && current.length > 0) {
        chunks.push(current.trim());
        current = current.slice(-chunkOverlap);
      }
      current += sentence;
    }
    if (current.trim()) {
      chunks.push(current.trim());
    }
    return chunks;
  }

  async handle(buffer: Buffer, jobKey: string, fileId: string): Promise<void> {
    this.logger.log(`Chunking & vectorizing markdown for job ${jobKey}`);

    const text = buffer.toString('utf-8');
    const chunks: string[] = this.segmentText(text);
    this.logger.log(`Split markdown into ${chunks.length} chunks`);
    await this.vectorSaver.saveTextChunksAsVectors(chunks, fileId, jobKey);
  }
}

File: packages/file-ingestor/src/handlers/markdown.handler.py
import logging
from langchain_text_splitters import RecursiveCharacterTextSplitter

supported_types = [
    "text/plain",
    "text/markdown",
    "application/octet-stream",
]
logger = logging.getLogger(__name__)
text_splitter = RecursiveCharacterTextSplitter(
    chunk_size=1000,
    chunk_overlap=200,
    length_function=len,
    add_start_index=True,
    separators=["\n\n", "\n", " ", ""],
)


def handle(buffer: bytes) -> str:
    text = buffer.decode("utf-8")
    if not text:
        logger.warning("Received empty buffer, returning empty string.")
        return ""

    all_chunks = text_splitter.split_text(text)
    if not all_chunks:
        logger.warning("No chunks were created from the text.")
        return ""

    logger.info(f"Created {len(all_chunks)} chunks from the text.")
    return all_chunks
